### PR TITLE
Faster cumsum and cumprod backwards

### PR DESCRIPTION
**Pull request by lezcano**
Piggybacking on https://github.com/pytorch/pytorch/pull/58747, now we can implement the backwards of `cumsum` and `cumprod` without tricks. This minimises the number of kernels that are launched in GPU, so we see a reasonable speed-up on GPU. We should also get a better stability for ill-conditioned inputs, as we do not perform any numerical tricks to get the result.

Note that the benchmarks test forward + backward, so the true speed-up on the backward should be even faster. Even more so in `cumsum`, as it requires less operations than the backward of `cumprod`.

<details>
<summary>
Test Script
</summary>

```python
from itertools import product

import torch
from torch.utils.benchmark import Compare, Timer



def get_timer(ndims, prod_dim, dim, num_threads, device):
    size = [500]*ndims
    size[dim] = prod_dim

    x = torch.rand(*size, device=device, requires_grad=True)
    # Make sure there are no zeros as the formula for the backward
    # that we are testing is for when the backward has no zeros
    with torch.no_grad():
        x.add_(1e-3)
    grad = torch.ones_like(x)

    timer = Timer(
        "torch.autograd.grad([x.cumprod(dim)], [x], grad_outputs=[grad])",
        globals={"x": x, "dim": dim, "grad": grad},
        label=f"Cumprod + Backwards {device}",
        description=f"dim: {dim}",
        sub_label=f"prod_dim: {prod_dim}",
        num_threads=num_threads,
    )

    return timer.blocked_autorange(min_run_time=5)


def get_params():
    ndims = 3
    dims = range(ndims)
    prod_dims = [10, 100, 500]
    for dim, prod_dim, device in product(dims, prod_dims, ("cpu", "cuda")):
        threads = (1, 2, 4) if device == "cpu" else (1,)
        for num_threads in threads:
            yield ndims, prod_dim, dim, num_threads, device


compare = Compare([get_timer(*params) for params in get_params()])
compare.trim_significant_figures()
compare.print()
```

</details>

<details>
<summary>
Benchmark PR
</summary>

```
[------------ Cumprod + Backwards cpu -------------]
                     |  dim: 0  |  dim: 1  |  dim: 2
1 threads: -----------------------------------------
      prod_dim: 10   |     11   |     14   |     12
      prod_dim: 100  |    260   |    270   |    260
      prod_dim: 500  |   1400   |   1550   |   1360
2 threads: -----------------------------------------
      prod_dim: 10   |      6   |      6   |      6
      prod_dim: 100  |    170   |    166   |    167
      prod_dim: 500  |    902   |    950   |    858
4 threads: -----------------------------------------
      prod_dim: 10   |      4   |      3   |      3
      prod_dim: 100  |    110   |    108   |    106
      prod_dim: 500  |    576   |    590   |    547

Times are in milliseconds (ms).

[------------ Cumprod + Backwards cuda ------------]
                     |  dim: 0  |  dim: 1  |  dim: 2
1 threads: -----------------------------------------
      prod_dim: 10   |    562   |    566   |   1075
      prod_dim: 100  |   5388   |   5394   |   6697
      prod_dim: 500  |  28170   |  27580   |  30740

Times are in microseconds (us).
```

</details>

<details>
<summary>
Benchmark master
</summary>

```
[------------ Cumprod + Backwards cpu -------------]
                     |  dim: 0  |  dim: 1  |  dim: 2
1 threads: -----------------------------------------
      prod_dim: 10   |     11   |     13   |     12
      prod_dim: 100  |    270   |    270   |    256
      prod_dim: 500  |   1500   |   1590   |   1300
2 threads: -----------------------------------------
      prod_dim: 10   |      6   |      6   |      6
      prod_dim: 100  |    170   |    170   |    164
      prod_dim: 500  |    911   |    940   |    840
4 threads: -----------------------------------------
      prod_dim: 10   |      4   |      4   |      4
      prod_dim: 100  |    111   |    109   |    105
      prod_dim: 500  |    570   |    590   |    536

Times are in milliseconds (ms).

[------------ Cumprod + Backwards cuda ------------]
                     |  dim: 0  |  dim: 1  |  dim: 2
1 threads: -----------------------------------------
      prod_dim: 10   |    616   |    597   |   1109
      prod_dim: 100  |   5976   |   5723   |   7017
      prod_dim: 500  |  31110   |  29160   |  32320

Times are in microseconds (us).
```

</details>

### CODE REVIEW

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -287,7 +284,7 @@ Tensor cumprod_backward(const Tensor& grad, const Tensor& input, int64_t dim, co
     return grad;
   }
   dim = at::maybe_wrap_dim(dim, input.dim());
-  const int64_t dim_size = input.size(dim);
+  const int64_t dim_size = input.sizes()[dim];
```
Comment (albanD) [inline aten/src/ATen/native/ReduceOps.cpp:287]: Why change that?

Comment (Lezcano) [reply aten/src/ATen/native/ReduceOps.cpp:287]: It's marginally faster when `dim >= 0`, and we are calling `wrap_dim` in the line before.

Comment (albanD) [reply aten/src/ATen/native/ReduceOps.cpp:287]: Why is it faster? It feels like it creates a temporary vector that you then index.
While the first one just get that number directly?

Comment (Lezcano) [reply aten/src/ATen/native/ReduceOps.cpp:287]: I think I read that `sizes()[dim]` was more performant in a comment in a PR or in the code, and I went on and looked into the code and convinced myself that that was the case. But now that I'm looking into it again, and the code does a `maybe_wrap_dim`, and then it calls into `size_at_unchecked`, which does
https://github.com/pytorch/pytorch/blob/bdb964f89f6ae2650ab73219cb28c49774b68547/c10/core/impl/SizesAndStrides.h#L210-L212
For some reason, I got confused and thought that this was also creating a small buffer, but it's just doing pointer arithmetic.

Sorry for that, reverting.

Comment (albanD) [reply aten/src/ATen/native/ReduceOps.cpp:287]: I may be wrong this this one!
cc @ngimel  might know better?

Comment (Lezcano) [reply aten/src/ATen/native/ReduceOps.cpp:287]: Actually, now I remember what was the reason. `sizes()` does not create a temporary vector, but an `ArrayRef` (note that we are not calling `.vec()`). As such, That is, in the worst case scenario, the copy of a pointer and a size:
https://github.com/pytorch/pytorch/blob/bdb964f89f6ae2650ab73219cb28c49774b68547/c10/util/ArrayRef.h#L76-L78
and then the indexing.

The indirection is done on an xvalue, and given that we have a pointer to const T, it should not be too difficult to be optimised away by the compiler.

On the other hand, it seems difficult to optimise away a number of if-elses that are in the `size(idx)`, in particular that of the `maybe_wrap_dim`.

All in all, the difference between each should be negligible, and perhaps readability would count more in this case, as this is not in a hotpath. That being said, it'd be nice to know what's the preferred way of doing this, or even having an `unsafe_size(size_t)`.

Comment (ngimel) [reply aten/src/ATen/native/ReduceOps.cpp:287]: @lezcano has provided a great explanation. In numbers:
```
In [5]:  timer = Timer("m1.size(0);", setup="at::Tensor m1=torch::empty({1,1,1});", language="c++", timer=timeit.default_timer)
   ...:  stats=timer.collect_callgrind(number=30, repeats=3)
   ...:  print(stats[1].as_standardized().stats(inclusive=False))
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7f95c87d9400>
  180  torch/include/ATen/core/TensorBody.h:main
  150  ???:c10::TensorImpl::sizes() const
  134  /tmp/benchmark_utils_jit_build__16 ... 814291730550202/timer_src.cpp:main
   60  torch/include/c10/core/WrapDimMinimal.h:main
   60  ???:c10::TensorImpl::dim() const
    4  torch/include/c10/util/irange.h:main

Total: 588

In [6]:  timer = Timer("m1.sizes()[0];", setup="at::Tensor m1=torch::empty({1,1,1});", language="c++", timer=timeit.default_timer)
   ...:  stats=timer.collect_callgrind(number=30, repeats=3)
   ...:  print(stats[1].as_standardized().stats(inclusive=False))
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7f95c885c3a0>
  150  ???:c10::TensorImpl::sizes() const
  105  /tmp/benchmark_utils_jit_build__16 ... 283917667944551/timer_src.cpp:main
   90  torch/include/ATen/core/TensorBody.h:main

Total: 345
```

**Review by albanD — APPROVED** [review]
LGTM!
